src/dfx/src/lib/builders/custom.rs

```diff
@@ -189,8 +189,16 @@ impl CanisterBuilder for CustomBuilder {
 
 fn run_command(args: Vec<String>, vars: &[super::Env<'_>]) -> DfxResult<()> {
     let (command_name, arguments) = args.split_first().unwrap();
-
-    let mut cmd = std::process::Command::new(command_name);
+    let command_path = Path::new(command_name);
```

Comment:
There are other places where this is broken, but dfx tries to let you execute it from anywhere within a project's directory tree.

See resolve_config_path 
 (near https://github.com/dfinity/sdk/blob/master/src/dfx/src/config/dfinity.rs#L491), which works its way up the directory tree until it finds dfx.json.

See also `Config::get_project_root()`, which returns the directory that contains dfx.json.  Using this would be in alignment with the ticket description
> just having "script.sh" as the custom builder command should run script.sh which lies just next to dfx.json.

Overall I'd suggest building an absolute path and trying to execute that, rather than prepending `./`.  What would be wrong with something like this?
```
    let command_path = if command_path.is_relative() {
        config.get_project_root().join(command_path)
    } else {
        command_path
    };
```

Reply:
I don't think this interacts with parent directories at all. As for an absolute path, that's fine if you say it's fine, I was just trying to change as little about the expected behavior as possible

Reply:
1. If we're going to assume that a filename refers to a file in a particular directory, check that it exists, and then execute it if so, then I think we should verify its existence and execute it by absolute path, rather than by prepending `./` to it. 

2. dfx is meant to work if run within any subdirectory of a project.  Though there are bugs that stop this from working, we don't have to add more today.  If we use `Config::get_project_root()` here, we don't introduce a new one.  This would also be consistent with this request from the ticket:

> just having "script.sh" as the custom builder command should run script.sh which lies just next to dfx.json.

Reply:
Oh, I see what you mean. So then should *any* relative script path be canonicalized against get_project_root before being passed to Command::new?

Reply:
I see what you mean, it's just weird to change the precise meaning of the path in the single use case. Anyway, if it's next to dfx.json then the original problem doesn't trigger from a subdirectory; if it is to be executable from a subdirectory then we would need to call Command::current_dir for the script, which requires canonicalizing *all* commands, not just in this special case. Should we do this?

Reply:
Yeah, I think we should canonicalize all commands as relative paths from the project root, and specify Command::current_dir as the project root.  In the common case (where dfx is being run from the project root), that will be like canonicalizing from the current directory, and using the current directory for execution.  What do you think?

Reply:
I think that's a great idea.